### PR TITLE
[MXNET-978] Higher order gradient for sigmoid

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
This PR supports higher order gradient for sigmoid operator.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by apeforest** [comment]
@kshitij12345 I have figured out how backward works when one of the inputs is an output of the forward node. Please review this PR. Thanks!

**Comment by apeforest** [comment]
@larroy @sxjscience Please help review this PR. Thanks!

**Comment by apeforest** [comment]
@larroy I also added the method to dump computation graph in imperative mode since it will be very useful for us to debug. However, it's still very rudimentary and we still need your help to implement a  
 more elegant way of printing out the graph info. thanks!

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 1]
```diff
@@ -121,7 +121,30 @@ The storage type of ``sigmoid`` output is always dense
...
+      // f''(x) = f'(x) * (1 - 2*f(x))
+      auto ones = MakeNode("ones_like", n->attrs.name + "_grad_ones", {n->inputs[1]}, nullptr, &n);
+      const std::unordered_map<std::string, std::string> args = {{"scalar", "2.0"}};
+      auto two_y = MakeNode("_mul_scalar", n->attrs.name + "_mul_two", {n->inputs[1]}, &args, &n);
+      auto one_minus_two_y = MakeNode("elemwise_sub", n->attrs.name + "_grad_sub",
+                                    {nnvm::NodeEntry{ones}, nnvm::NodeEntry{two_y}}, nullptr, &n);
+      auto grad_grad_mid = MakeNode("elemwise_mul", n->attrs.name + "_grad_mul",
+                                    {n->inputs[0], nnvm::NodeEntry{one_minus_two_y}}, nullptr, &n);
+      // when building gradient graph, the backward node of n->inputs[1] will be
+      // added to the graph again, therefore f`(x) will be multiplied
```
Comment (kshitij12345) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Doesn't this behaviour seem a bit odd? Is this actually the expected behaviour? What would have happened if this was `split` like function where we would have had many outputs, backward of all outputs be added? Actually I am confused by the behaviour.

Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Yes, this is actually the expected behavior. The nnvm graph will perform a DFS traverse when performing backward pass. Since the input to this backward_sigmoid node is an output of another node sigmoid, during the backward pass the gradient function sigmoid will be invoked.

See:

This is to collect dependent nodes in the graph during RecordOp:
https://github.com/apache/incubator-mxnet/blob/master/src/imperative/imperative.cc#L180

This is to actually perform the backward pass:
https://github.com/dmlc/tvm/blob/21935dcbf56ad3bd66ebff9891a6bc3865b8106d/nnvm/src/pass/gradient.cc#L126

https://github.com/dmlc/tvm/blob/21935dcbf56ad3bd66ebff9891a6bc3865b8106d/nnvm/src/pass/gradient.cc#L190

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 2]
```diff
@@ -121,7 +121,30 @@ The storage type of ``sigmoid`` output is always dense
...
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_sigmoid,
-                                               unary_bwd<mshadow_op::sigmoid_grad>);
+                                               unary_bwd<mshadow_op::sigmoid_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+    [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+      // n->inputs[0] : y_grad
+      // n->inputs[1] : f(x) = sigmoid(x)
+      // ograds[0] : head_grads
+      // f''(x) = f'(x) * (1 - 2*f(x))
```
Comment (larroy) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: ok

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (larroy) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: I also didn't get that. A drawing might help. I would expect that you multiply with the backward node itself "n"

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 2 above)
Comment (larroy) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: ok

File: tests/python/unittest/test_higher_order_grad.py [hunk 3]
```diff
@@ -106,6 +106,23 @@ def grad_grad_op(x):
         check_second_order_unary(array, log10, grad_grad_op)
 
 
+@with_seed()
+def test_sigmoid():
```
Comment (larroy) [inline tests/python/unittest/test_higher_order_grad.py]: looks correct to me.

File: src/imperative/imperative.cc [hunk 4]
```diff
@@ -501,6 +501,10 @@ std::vector<NDArray*> Imperative::Backward(
     }
   }
 
+  if (dmlc::GetEnv("MXNET_MEM_PLAN_VERBOSE_LOGGING", false)) {
```
Comment (larroy) [inline src/imperative/imperative.cc]: nice hack :)

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: I guess that is what he meant as 
```
      // n->inputs[0] : y_grad
      // n->inputs[1] : f(x) = sigmoid(x)
      // ograds[0] : head_grads
      // f''(x) = f'(x) * (1 - 2*f(x))
```
Backward of node `n->inputs[1]` is the Node `n` itself.

One up for the visual drawing/graph of what is actually happening.

Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: IMG-3494

Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Here is the entire computation graph including forward pass and the second order backward pass. You can dump out the text version using 
`export MXNET_MEM_PLAN_VERBOSE_LOGGING=1`

Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: > Backward of node `n->inputs[1]` is the Node `n` itself.

`n` here is the backward node _backward_sigmoid.
n->inputs[1] is the forward node sigmoid.
During the second `autograd.grad()` function, it will actually trigger the FGradient function of `n->inputs[1]` which is another _backward_sigmoid node in addition to `n`. So that is why you will see two `_back_sig` nodes shown in the computation graph above.

Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: It will be easier to understand if you modify the FGradient function of _backward_sigmoid with only ograd[0] as input and dump out the graph. It will help you understand better.

```suggestion
      std::vector<nnvm::NodeEntry> ret;
      ret.emplace_back(ograds[0]);  // this output is not passed out if gradient w.r.t x only
      ret.emplace_back(ograds[0]);  // instead of passing back a complicated function, simply pass back the head_grad
```

File: src/imperative/imperative.cc (hunk 4 above)
Comment (access2rohit) [reply src/imperative/imperative.cc]: Can you explain how logging of static memory helps here ?

Comment (larroy) [reply src/imperative/imperative.cc]: He wants to dump the graph. maybe should be separate PR?

Comment (apeforest) [reply src/imperative/imperative.cc]: It's actually dumping out the computation graph.

Comment (larroy) [reply src/imperative/imperative.cc]: @apeforest isn't that what I wrote?

Comment (apeforest) [reply src/imperative/imperative.cc]: @larroy I did not see your comment before I refresh the page.  Our minds think alike :)
Sure, I can separate this into a different PR. I added here only to help @kshitij12345 dump out the graph and understand the backward pass better.

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (larroy) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Thanks, I got it. Thanks for the explanation in person ;-)

Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Thanks I'll look at it today.

File: src/imperative/imperative.cc (hunk 4 above)
Comment (larroy) [reply src/imperative/imperative.cc]: The PR is great, up to you to separate not a big deal. I'm not a radical on my reviews.

Comment (kshitij12345) [reply src/imperative/imperative.cc]: @apeforest Really like the dump, its great help to actually see the graph. Thank You.

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: I kind of see from the graph you have drawn and dumped graph as to what is happening. I still strongly that it is the incorrect behaviour. 

To check that I modified the test code slightly to also check for the 3rd order gradient as it should come for free since the second order are composed from differentiable functions.

Have tested for `sin`, `cos`, `log`, `sigmoid`. Out of them only `sigmoid` fails for the third order.

I have also attached the hand computation of third-order for `sigmoid`, so please verify that just to make sure that it is not incorrect.

For some reason it is rotated.
3_rd_ord


```python

import math
from mxnet import nd, autograd
from mxnet.test_utils import assert_almost_equal, random_arrays, rand_shape_nd
from common import with_seed


@with_seed()
def test_sin():
    def sin(x):
        return nd.sin(x)

    def grad_grad_op(x):
        return -nd.sin(x)
    
    def grad_grad_grad_op(x):
        return -nd.cos(x)

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, sin, grad_grad_op, grad_grad_grad_op)


@with_seed()
def test_cos():
    def cos(x):
        return nd.cos(x)

    def grad_grad_op(x):
        return -nd.cos(x)
    
    def grad_grad_grad_op(x):
        return nd.sin(x)

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, cos, grad_grad_op, grad_grad_grad_op)

@with_seed()
def test_log():
    def log(x):
        return nd.log(x)

    def grad_grad_op(x):
        return -1/(x**2)

    def grad_grad_grad_op(x):
        return 2/(x**3)

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, log, grad_grad_op, grad_grad_grad_op)


@with_seed()
def test_sigmoid():
    def sigmoid(x):
        return nd.sigmoid(x)

    def grad_op(x):
        return sigmoid(x) * (1 - sigmoid(x))

    def grad_grad_op(x):
        return grad_op(x) * (1 - 2 * sigmoid(x))

    def grad_grad_grad_op(x):
        return grad_grad_op(x) - 2 * ( grad_op(x)**2 + grad_grad_op(x) * sigmoid(x))

    for dim in range(1, 5):
        shape = rand_shape_nd(dim)
        array = random_arrays(shape)
        check_second_order_unary(array, sigmoid, grad_grad_op, grad_grad_grad_op)


def check_second_order_unary(x, op, grad_grad_op, grad_grad_grad_op):
    x = nd.array(x)
    grad_grad_x = grad_grad_op(x)
    grad_grad_grad_x = grad_grad_grad_op(x)
    x.attach_grad()

    # Manual head_grads.
    y_grad = nd.random.normal(shape=x.shape)
    head_grad_grads = nd.random.normal(shape=x.shape)

    # Perform compute.
    with autograd.record():
        y = op(x)
        x_grad = autograd.grad(heads=y, variables=x, head_grads=y_grad,
                               create_graph=True, retain_graph=True)[0]
        
        x_grad_grad = autograd.grad(heads=x_grad, variables=x, head_grads=head_grad_grads,
                                create_graph=True, retain_graph=True)[0]
    x_grad_grad.backward()

    # Compute expected values.
    expected_grad_grad = grad_grad_x.asnumpy() * head_grad_grads.asnumpy() * \
        y_grad.asnumpy()

    expected_grad_grad_grad = grad_grad_grad_x.asnumpy() * head_grad_grads.asnumpy() * \
        y_grad.asnumpy()

    # Validate the gradients.
    assert_almost_equal(expected_grad_grad, x_grad_grad.asnumpy())
    assert_almost_equal(expected_grad_grad_grad, x.grad.asnumpy())

if __name__ == '__main__':
    import nose
    nose.runmodule()
```

Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: I suspect the discrepancy in 3rd order gradient is not because an error in my implementation of grad_grad_input, but because I did not return back the first grad_grad in correctly. After all, the first output may be useful in calculating higher order even if they are not visible output at Python level. I will look into it.

**Comment by larroy** [comment]
Can we merge this?

**Comment by apeforest** [comment]
I verified the result is the same as pytorch

```
import torch
import numpy as np
import math

op = lambda x: torch.sigmoid(x)
grad_op = lambda x: op(x) * (1 - op(x))
grad_grad_op = lambda x: grad_op(x) * (1 - 2 * op(x))
grad_grad_grad_op = lambda x: grad_grad_op(x) - 2 * ( grad_op(x)**2 + grad_grad_op(x) * op(x))

x = torch.tensor(np.array([1, 2, 3]), dtype=torch.float32)
head_grads = torch.tensor(np.array([1, 1, 1]), dtype=torch.float32) * 0.5
head_grad_grads = torch.tensor(np.array([1, 1, 1]), dtype=torch.float32) * 0.6
head_grad_grad_grads = torch.tensor(np.array([1, 1, 1]), dtype=torch.float32) * 0.7
x.requires_grad = True
head_grads.requires_grad = True

y = op(x)
x_grad = torch.autograd.grad(y, x, grad_outputs= head_grads, create_graph=True, retain_graph=True)[0]
expected_grad_x = (grad_op(x) * head_grads).detach().numpy()
print('expected_grad_x = {}'.format(expected_grad_x))
print('grad_x          = {}'.format(x_grad.detach().numpy()))
x_grad_grad = torch.autograd.grad(x_grad, x, grad_outputs= head_grad_grads, create_graph=True, retain_graph=True)[0]
x_grad_grad.backward(head_grad_grad_grads)

expected_grad_grad_x = (grad_grad_op(x) * head_grads * head_grad_grads).detach().numpy()
expected_head_grad = (grad_op(x) * head_grad_grads).detach().numpy()
expected_grad_grad_grad_x = (grad_grad_grad_op(x) * head_grads * head_grad_grads * head_grad_grad_grads).detach().numpy()

print('expected_grad_grad_x = {}'.format(expected_grad_grad_x))
print('grad_grad_x          = {}'.format(x_grad_grad.detach().numpy()))
print('expected_grad_grad_grad_x = {}'.format(expected_grad_grad_grad_x))
print('grad_grad_grad_x          = {}'.format(x.grad.detach().numpy()))
```

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (apeforest) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: @kshitij12345 I have fixed the issue. The result can pass your test now. Please review again. Thanks!

**Comment by apeforest** [comment]
@sxjscience Please help to review this PR. Thanks!

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: LGTM.

**Comment by apeforest** [comment]
@kshitij12345 could you approve the PR if everything looks good to you now? thx
